formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/schema/ProtoBufSchemaGenerator.kt

```diff
@@ -0,0 +1,509 @@
+@file:OptIn(ExperimentalSerializationApi::class)
+
+package kotlinx.serialization.protobuf.schema
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.protobuf.*
+
+/**
+ * Experimental generator of ProtoBuf schema that is compatible with [serializable][Serializable] Kotlin classes
+ * and data encoded and decoded by [ProtoBuf] format.
+ *
+ * The schema is generated based on provided [SerialDescriptor] and is compatible with proto2 schema definition.
+ * An arbitrary Kotlin class represent much wider object domain than the ProtoBuf specification, thus schema generator
+ * has the following list of restrictions:
+ *
+ *  * Serial name of the class and all its fields should be a valid Proto [identifier](https://developers.google.com/protocol-buffers/docs/reference/proto2-spec)
+ *  * Nullable values are allowed only for Kotlin [nullable][SerialDescriptor.isNullable] types
+ *  * [SerialName] of the type uniquely identifies the proto message type and two or more fields with the same serial name
+ *    are considered to have the same type.
+ *  * Nested collections, e.g. `List<List<Int>>` are represented using the artificial wrapper message in order to distinguish
+ *    repeated fields boundaries.
+ *  * Default Kotlin values are not representable in proto schema. A special commentary is generated for properties with default values.
+ *  * Empty nullable collections are not supported by the generated schema and will be prohibited in [ProtoBuf] as well
+ *    due to their ambiguous nature.
+ *
+ * Temporary limitation:
+ *  * [Contextual] data is represented as as `bytes` type
+ *  * [Polymorphic] data is represented as a artificial `KotlinxSerializationPolymorphic` message.
+ *
+ * Other types are mapped according to their specification: primitives as primitives, lists as 'repeated' fields and
+ * maps as 'repeated' map entries.
```

Comment:
For some cases, Map is a `map`, for others is a `repeated`, it depends on the type of key.